### PR TITLE
Add normals/nodes to dcoll

### PR DESCRIPTION
**Pull request by thomasgibson**
cc @MTCam @inducer

For discussion based on this thread: https://github.com/illinois-ceesd/mirgecom/pull/360#discussion_r638097941

### CODE REVIEW

File: grudge/discretization.py [hunk 1]
```diff
@@ -519,6 +521,38 @@ def order(self):
...
+        :returns: an object array of :class:`~meshmode.dof_array.DOFArray`\ s
+        """
+        if dd is None:
+            return self._volume_discr.nodes()
+        else:
+            return self.discr_from_dd(dd).nodes()
+
+
+    @memoize_method
+    def normal(self, dd):
```
Comment (MTCam) [inline grudge/discretization.py:540]: Will the memoization here do anything to prevent update if the mesh moves?  Just curious.

**Comment by thomasgibson** [comment]
I am leaning towards delaying https://github.com/illinois-ceesd/mirgecom/pull/360 until https://github.com/inducer/grudge/pull/74 in merged in Grudge. The reason being is that this will be (hopefully soon) the right place to start.

File: grudge/discretization.py (hunk 1 above)
Comment (thomasgibson) [reply grudge/discretization.py:540]: That's a good question. I don't think so. This would have to be inferred I guess if the `nodes` are changing.

Comment (inducer) [reply grudge/discretization.py:540]: See https://github.com/inducer/meshmode/issues/123 for some related discussion. IMO, generally the safest approach would be to discard and rebuild the whole collection when it changes, while working to make that a (reasonably) cheap operation.

File: grudge/discretization.py [hunk 2]
```diff
@@ -519,6 +521,38 @@ def order(self):
         from pytools import single_valued
         return single_valued(egrp.order for egrp in self._volume_discr.groups)
 
+    # {{{ Discretization-specific geometric properties
+
+    def nodes(self, dd=None):
```
Comment (inducer) [inline grudge/discretization.py:526]: Add tests?

**Comment by inducer** [comment]
I agree that this should wait until after #74.

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

File: grudge/discretization.py (hunk 1 above)
Comment (MTCam) [reply grudge/discretization.py:540]: Will this discarding and rebuilding be ok when the structures surface regresses at every substep?  Even in the absence of regression (caused by erosion, or burning); if we have real coupling there will be deformation displacements - which definitely occur at the substep level; that is, every time the surface load (i.e. the fluid pressure) changes.  This approach seems like a lot of tear-down-setup overhead.

Comment (inducer) [reply grudge/discretization.py:625]: The thing is that making new objects is (relatively) cheap. If it's a new mesh, then by far the most expensive thing will be remapping to a new mesh. If it's smooth deformations, then IMO the cost will be dominated by updating/recomputing area elements and mapping derivatives. Making a new `DiscretizationCollection`, by comparison, is dirt cheap.

Comment (inducer) [reply grudge/discretization.py:625]: [That constructor](https://github.com/inducer/grudge/blob/0a7b2fb2e4e170809b55e0c3c193ae1dbc4a61a4/grudge/discretization.py#L74-L162) currently does a bit more work than it needs to, but I think that can be addressed. #104

File: grudge/discretization.py [hunk 3]
```diff
@@ -601,6 +608,33 @@ def order(self):
...
+        r"""Return the nodes of a discretization specified by *dd*.
+
+        :arg dd: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one.
+            Defaults to the base volume discretization.
+        :returns: an object array of :class:`~meshmode.dof_array.DOFArray`\ s
```
Comment (inducer) [inline grudge/discretization.py:618]:
```suggestion
        :returns: an object array of frozen :class:`~meshmode.dof_array.DOFArray`\ s
```

File: grudge/discretization.py [hunk 4]
```diff
@@ -601,6 +608,33 @@ def order(self):
...
+        if dd is None:
+            dd = DD_VOLUME
+        return self.discr_from_dd(dd).nodes()
+
+    @memoize_method
+    def normal(self, dd):
+        r"""Get the unit normal to the specified surface discretization, *dd*.
+
+        :arg dd: a :class:`~grudge.dof_desc.DOFDesc` as the surface discretization.
+        :returns: an object array of :class:`~meshmode.dof_array.DOFArray`\ s.
```
Comment (inducer) [inline grudge/discretization.py:629]:
```suggestion
        :returns: an object array of frozen :class:`~meshmode.dof_array.DOFArray`\ s.
```

**Review by inducer — APPROVED** [review]
Thanks for working on this! LGTM, I'll apply these.

**Pull request merged**